v3/histogram_test.go

```diff
@@ -0,0 +1,263 @@
+package v3
+
+import (
+	"go/constant"
+	"reflect"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
+	"bytes"
+)
+
+type testBucket struct {
+	numEq      int64
+	numRange   int64
+	upperBound int64
+}
+
+type testHistogram struct {
+	rowCount      int64
+	distinctCount int64
+	nullCount     int64
+	lowerBound    int64
+	buckets       []testBucket
+}
+
+func histogramFromTestHistogram(t *testing.T, histograms []testHistogram) []histogram {
+	out := make([]histogram, len(histograms))
+	for i, h := range histograms {
+		buckets := make([]bucket, len(h.buckets))
+		for j, b := range h.buckets {
+			v := &tree.NumVal{Value: constant.MakeInt64(b.upperBound)}
+			ub, err := v.ResolveAsType(nil, types.Int)
+			if err != nil {
+				t.Fatalf("could not create upper bound Datum: %s: %v", v, err)
+			}
+
+			buckets[j] = bucket{
+				numEq:      b.numEq,
+				numRange:   b.numRange,
+				upperBound: ub,
+			}
+		}
+
+		v := &tree.NumVal{Value: constant.MakeInt64(h.lowerBound)}
+		lb, err := v.ResolveAsType(nil, types.Int)
+		if err != nil {
+			t.Fatalf("could not create lower bound Datum: %s: %v", v, err)
+		}
+
+		out[i] = histogram{
+			rowCount:      h.rowCount,
+			distinctCount: h.distinctCount,
+			nullCount:     h.nullCount,
+			lowerBound:    lb,
+			buckets:       buckets,
+		}
+	}
+
+	return out
+}
+
+func TestHistogram(t *testing.T) {
+	testHistograms := histogramFromTestHistogram(t, []testHistogram{
```

Comment:
We should figure out a way to fit this into the `logic_test` framework. For example, we could have something like:

```
exec
SELECT * FROM histogram.a.x WHERE x > 4
```

Notice that this is an `exec` vs the other logic test commands. A bit more work on the test infrastructure side, but I think it would do a lot for the test clarity. The output would be the filtered histogram (i.e. `histogram.String()`).

Reply:
PS I can help with this if you'd like.